plugins/lookup/merge_variables.py

```diff
@@ -131,22 +139,52 @@ def _verify_and_get_type(variable):
 
 
 class LookupModule(LookupBase):
-
     def run(self, terms, variables=None, **kwargs):
         self.set_options(direct=kwargs)
         initial_value = self.get_option("initial_value", None)
         self._override = self.get_option('override', 'error')
         self._pattern_type = self.get_option('pattern_type', 'regex')
+        self._groups = self.get_option('groups', None)
 
         ret = []
         for term in terms:
             if not isinstance(term, str):
                 raise AnsibleError("Non-string type '{0}' passed, only 'str' types are allowed!".format(type(term)))
 
-            ret.append(self._merge_vars(term, initial_value, variables))
+            if not self._groups:  # consider only own variables
+                ret.append(self._merge_vars(term, initial_value, variables))
+            else:  # consider variables of hosts in given groups
+                cross_host_merge_result = None
+                cross_host_var_type = ""
+                for host in variables["hostvars"]:
+                    if self._is_host_in_allowed_groups(variables["hostvars"][host]["group_names"]):
+                        host_result = self._merge_vars(term, initial_value, variables["hostvars"][host])
+
+                        if cross_host_merge_result is None:
+                            if host_result:
+                                cross_host_merge_result = host_result
+                                cross_host_var_type = _verify_and_get_type(cross_host_merge_result)
+                        else:
+                            if host_result:
+                                if cross_host_var_type == "dict":
+                                    cross_host_merge_result = self._merge_dict(host_result, cross_host_merge_result, [""])
+                                else:
+                                    cross_host_merge_result += host_result
+
+                ret.append(cross_host_merge_result)
```

Comment:
```suggestion
            else:  # consider variables of hosts in given groups
                cross_host_merge_result = initial_value
                for host in variables["hostvars"]:
                    if self._is_host_in_allowed_groups(variables["hostvars"][host]["group_names"]):
                        cross_host_merge_result = self._merge_vars(term, cross_host_merge_result, variables["hostvars"][host])

                ret.append(cross_host_merge_result)
```

By using the `initial_value` of the `_merge_vars` no duplication of the merge functionality is needed.
It looks like this evens solves a bug, see my comment for the integration test.

Reply:
Thanks, that a nice improvement! I didn't like the duplication in the first place, but I didn't think of just reusing  intermediate is enough here.